bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+                    bool found_type = offsetof_to_type.contains(op),
+                        found_member = offsetof_to_member.contains(op),
+                        is_legal = !(found_type && found_member);
+                    return is_legal;
+                } else if (is_container_of) {
+                    exp.getExpansion().walk([&](Op *cur) {
+                        if (auto param = mlir::dyn_cast<MP>(cur)) {
+                            auto param_name = param.getParameterName();
+                            if ("ptr" == param_name) {
+                                container_of_to_ptr.insert({ op, cur });
+                            }
+                        } else if (auto mem_op = mlir::dyn_cast<RMO>(cur)) {
+                            mlir::Value record = mem_op.getRecord();
+                            mlir::Type record_ty = record.getType();
+                            if (auto pty = mlir::dyn_cast<PT>(record_ty)) {
+                                mlir::Type elem_ty = pty.getElementType();
+                                TA ty_attr = TA::get(elem_ty);
+                                container_of_to_type.insert({ op, ty_attr });
+                            } else {
+                                TA ty_attr = TA::get(record_ty);
+                                container_of_to_type.insert({ op, ty_attr });
+                            }
+                            mlir::StringAttr mem_name = mem_op.getNameAttr();
+                            container_of_to_member.insert({ op, mem_name });
+                        }});
+                    bool found_ptr = container_of_to_ptr.contains(op),
+                        found_type = container_of_to_type.contains(op),
+                        found_member = container_of_to_member.contains(op),
+                        is_legal = !(found_ptr && found_type && found_member);
+                    return is_legal;
+                } else if (is_rcu_dereference) {
+                    exp.getExpansion().walk([&](Op *cur) {
+                        if (auto param = mlir::dyn_cast<MP>(cur)) {
+                            auto param_name = param.getParameterName();
+                            if (param_name == "p") {
+                                rcu_dereference_to_p.insert({ op, cur });
+                            }
+                        }});
+                    bool found_p = rcu_dereference_to_p.contains(op),
+                        is_legal = !found_p;
+                    return is_legal;
+                } else if (is_smp_mb) {
+                    return false;
+                } else {
+                    return true;
+                }
+            });
+        trg.addDynamicallyLegalOp<vast::hl::ForOp>(
+            [](Op *op) {
+                auto for_op = mlir::dyn_cast<vast::hl::ForOp>(op);
+                using MP = macroni::macroni::MacroParameter;
+                for_op.getCondRegion().walk(
+                    [&](Op *op2) {
```

Comment:
Nested lambdas are kinda hard to parse. I would extract them as local vars. And then you can use proper names instead of `op2`. You almost never want to use digits as suffices to make new names.